src/backend/crossterm.rs

```diff
@@ -36,149 +155,96 @@ impl Backend {
-
-                    Event::Mouse {
-                        event,
-                        position,
-                        offset: Vec2::zero(),
+                    CMouseEvent::Up(_, x, y, _) => {
+                        event = MouseEvent::Release(self.last_button.unwrap());
+                        position = (x, y).into();
                     }
-                }
-                CMouseEvent::Release(x, y) if self.last_button.is_some() => {
-                    let event = MouseEvent::Release(self.last_button.unwrap());
-                    let position = (x, y).into();
-
-                    Event::Mouse {
-                        event,
-                        position,
-                        offset: Vec2::zero(),
+                    CMouseEvent::Drag(_, x, y, _) => {
+                        event = MouseEvent::Hold(self.last_button.unwrap());
+                        position = (x, y).into();
                     }
-                }
-                CMouseEvent::Hold(x, y) if self.last_button.is_some() => {
-                    let event = MouseEvent::Hold(self.last_button.unwrap());
-                    let position = (x, y).into();
-
-                    Event::Mouse {
-                        event,
-                        position,
-                        offset: Vec2::zero(),
+                    CMouseEvent::ScrollDown(x, y, _) => {
+                        event = MouseEvent::WheelDown;
+                        position = (x, y).into();
                     }
+                    CMouseEvent::ScrollUp(x, y, _) => {
+                        event = MouseEvent::WheelDown;
+                        position = (x, y).into();
+                    }
+                };
+
+                Event::Mouse {
+                    event,
+                    position,
+                    offset: Vec2::zero(),
                 }
-                _ => {
-                    log::warn!(
-                        "Unknown mouse button event {:?}!",
-                        mouse_event
-                    );
-                    Event::Unknown(vec![])
-                }
-            },
-            _ => {
-                log::warn!("Unknown mouse event {:?}!", event);
-                Event::Unknown(vec![])
             }
+            CEvent::Resize(_, _) => Event::WindowResize,
         }
     }
 }
 
 impl backend::Backend for Backend {
-    fn name(&self) -> &str {
-        "crossterm"
-    }
-
     fn poll_event(&mut self) -> Option<Event> {
-        self.async_reader.next().map(|event| self.map_key(event))
+        match poll(Duration::from_millis(0)) {
+            Ok(true) => match read() {
+                Ok(event) => Some(self.map_key(event)),
+                Err(_) => None,
+            },
+            _ => None,
+        }
     }
 
     fn finish(&mut self) {
-        input().disable_mouse_mode().unwrap();
-        cursor().show().unwrap();
+        execute!(self.stdout_mut(), DisableMouseCapture, Show)
```

Comment:
Currently exiting an application crashes here: 
```
thread 'main' panicked at 'already borrowed: BorrowMutError', src/libcore/result.rs:1187:5
stack backtrace:
   0: backtrace::backtrace::libunwind::trace
             at /cargo/registry/src/github.com-1ecc6299db9ec823/backtrace-0.3.40/src/backtrace/libunwind.rs:88
   1: backtrace::backtrace::trace_unsynchronized
             at /cargo/registry/src/github.com-1ecc6299db9ec823/backtrace-0.3.40/src/backtrace/mod.rs:66
   2: std::sys_common::backtrace::_print_fmt
             at src/libstd/sys_common/backtrace.rs:84
   3: <std::sys_common::backtrace::_print::DisplayBacktrace as core::fmt::Display>::fmt
             at src/libstd/sys_common/backtrace.rs:61
   4: core::fmt::write
             at src/libcore/fmt/mod.rs:1030
   5: std::io::Write::write_fmt
             at src/libstd/io/mod.rs:1412
   6: std::sys_common::backtrace::_print
             at src/libstd/sys_common/backtrace.rs:65
   7: std::sys_common::backtrace::print
             at src/libstd/sys_common/backtrace.rs:50
   8: std::panicking::default_hook::{{closure}}
             at src/libstd/panicking.rs:188
   9: std::panicking::default_hook
             at src/libstd/panicking.rs:205
  10: std::panicking::rust_panic_with_hook
             at src/libstd/panicking.rs:464
  11: std::panicking::continue_panic_fmt
             at src/libstd/panicking.rs:373
  12: rust_begin_unwind
             at src/libstd/panicking.rs:302
  13: core::panicking::panic_fmt
             at src/libcore/panicking.rs:82
  14: core::result::unwrap_failed
             at src/libcore/result.rs:1187
  15: core::result::Result<T,E>::expect
             at /rustc/5c5b8afd80e6fa1d24632153cb2257c686041d41/src/libcore/result.rs:982
  16: core::cell::RefCell<T>::borrow_mut
             at /rustc/5c5b8afd80e6fa1d24632153cb2257c686041d41/src/libcore/cell.rs:876
  17: cursive::backend::crossterm::Backend::stdout_mut
             at src/backend/crossterm.rs:180
  18: <cursive::backend::crossterm::Backend as cursive::backend::Backend>::finish::{{closure}}
             at src/backend/crossterm.rs:242
  19: core::result::Result<T,E>::and_then
             at /rustc/5c5b8afd80e6fa1d24632153cb2257c686041d41/src/libcore/result.rs:713
  20: <cursive::backend::crossterm::Backend as cursive::backend::Backend>::finish
             at src/backend/crossterm.rs:242
```

Reply:
Note that since we have `&mut self` here, we can directly use `RefCell::get_mut`:
```suggestion
        execute!(self.stdout.get_mut(), DisableMouseCapture, Show)
```

Reply:
That is strange. Should not happen. I haven't yet debugged on Linux btw.

Reply:
Found the problem: in `write_cout!`, you use `$write` twice:
https://github.com/crossterm-rs/crossterm/blob/master/src/utils/macros.rs#L20

The macro expands `$write` to the _expression_ `self.stdout_mut()`, not to the result. So it ends up calling `stdout_mut()` twice in there, which obviously fails.

Reply:
I did a macro rewrite in this PR, does switching to this branch fixes this? 

https://github.com/crossterm-rs/crossterm/pull/326